File: 01_The_Tyranny_of_the_Rocket_Equation/solution.js
const fs = require('fs')

const puzzle_input = fs.readFileSync('input.txt').toString().split(
    "\n").filter(Boolean).map(Number)

const calcTotalFuel = (input, option) =>
    input.reduce((total, moduleMass) => {
        const fuelNeeded = option == 2 ? massFuel(moduleMass) : requiredFuel(moduleMass)
        return total + fuelNeeded
    }, 0)

const requiredFuel = (mass) => Math.floor(mass / 3) - 2

const massFuel = (mass) => {
    const fuel = requiredFuel(mass)
    return fuel < 0 ? 0 : fuel + massFuel(fuel)
}

console.log('Part 1: ', calcTotalFuel(puzzle_input))
console.log('Part 2: ', calcTotalFuel(puzzle_input, 2))

File: 01_The_Tyranny_of_the_Rocket_Equation/solution.py
from functools import reduce


with open('input.txt') as f:
    puzzle_input = list(map(int, f.read().splitlines()))


def required_fuel(mass):
    return int(mass / 3) - 2


def mass_fuel(mass):
    fuel = required_fuel(mass)
    return fuel + mass_fuel(fuel) if fuel >= 0 else 0


def calc_total_fuel(array, option):
    def accumulator(total, module_mass):
        fuel_needed = required_fuel(module_mass)
        if option == 2:
            fuel_needed = mass_fuel(module_mass)
        return total + fuel_needed
    return reduce(accumulator, array, 0)


print('Part 1:', calc_total_fuel(puzzle_input, 1))
print('Part 2:', calc_total_fuel(puzzle_input, 2))
